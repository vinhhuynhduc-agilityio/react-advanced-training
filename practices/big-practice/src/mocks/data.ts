export const mockUsers = [
  {
    id: "d290f1ee-6c54-4b01-90e6-d701748f0851",
    fullName: "Joe Bloggs",
    earnings: "$11500",
    email: "john@example.com",
    avatarUrl: "https://i.pravatar.cc/150?img=1",
    registered: "May 21, 2020 17:02:06",
    lastUpdated: "Nov 11, 2024 14:34:21",
  },
  {
    id: "f47ac10b-58cc-4372-a567-0e02b2c3d479",
    fullName: "Jane Smith",
    earnings: "$36911",
    email: "jane@example.com",
    avatarUrl: "https://i.pravatar.cc/150?img=2",
    registered: "June 10, 2021 09:20:15",
    lastUpdated: "Nov 18, 2024 17:33:37",
  },
  {
    id: "123e4567-e89b-12d3-a456-426614174000",
    fullName: "Alice Brown",
    earnings: "$7200",
    email: "alice@example.com",
    avatarUrl: "https://i.pravatar.cc/150?img=3",
    registered: "April 15, 2020 14:02:06",
    lastUpdated: "October 11, 2023 10:00:00",
  },
];

export const mockProject = [
  {
    id: "f2d32cb6-12c7-4ae7-bb28-74f16d1d2cbb",
    projectName: "Support",
  },
  {
    id: "c1c20a2e-f450-4875-96e8-334f92e9b3b5",
    projectName: "Failure Testing",
  },
  {
    id: "e38e56a2-4d3c-469d-8345-45d6b5fae9f9",
    projectName: "Quality Management",
  },
  {
    id: "f49b68f4-5e5b-457d-bf16-1b5b78e6f5f1",
    projectName: "Data Quality",
  },
  {
    id: "47c80c97-6575-45eb-a25b-75ee485bf9fb",
    projectName: "Test",
  },
];

export const mockTasks = [
  {
    id: "71e564f4-7c18-47f7-89f2-abe4b7ec2854",
    userId: "d290f1ee-6c54-4b01-90e6-d701748f0851",
    projectId: "f2d32cb6-12c7-4ae7-bb28-74f16d1d2cbb",
    taskName: "Build test",
    startDate: "10 Aug 24",
    completedDate: "15 Nov 24",
    currency: 2000,
    status: true,
    projectName: "Support",
    fullName: "Joe Bloggs",
  },
  {
    id: "82e564f4-7c18-47f7-89f2-abe4b7ec2856",
    userId: "d290f1ee-6c54-4b01-90e6-d701748f0851",
    projectId: "c1c20a2e-f450-4875-96e8-334f92e9b3b5",
    taskName: "Build Frontend",
    startDate: "05 Sep 23",
    completedDate: "27 Oct 24",
    currency: 4000,
    status: true,
    projectName: "Failure Testing",
    fullName: "Joe Bloggs",
  },
  {
    id: "u2e564f4-7c18-47f7-89f2-abe4b7ec2878",
    userId: "f47ac10b-58cc-4372-a567-0e02b2c3d479",
    projectId: "c1c20a2e-f450-4875-96e8-334f92e9b3b5",
    taskName: "Develop Chatbot Feature",
    startDate: "23 Sep 23",
    completedDate: "24 Oct 24",
    currency: 3200,
    status: true,
    projectName: "Failure Testing",
    fullName: "Jane Smith",
  },
  {
    id: "7ce698fc-c0ee-4d8c-8a1b-df988a96c22f",
    userId: "f47ac10b-58cc-4372-a567-0e02b2c3d479",
    projectId: "47c80c97-6575-45eb-a25b-75ee485bf9fb",
    taskName: "test today",
    startDate: "13 Nov 24",
    completedDate: "incomplete",
    currency: 1111,
    status: false,
    projectName: "Test",
    fullName: "Jane Smith",
  },
  {
    id: "r0s1t2u3-7c18-47f7-89f2-abe4b7ec2949",
    userId: "123e4567-e89b-12d3-a456-426614174000",
    projectId: "47c80c97-6575-45eb-a25b-75ee485bf9fb",
    taskName: "Data Migration",
    startDate: "15 Aug 24",
    completedDate: "20 Feb 24",
    currency: 2800,
    status: true,
    projectName: "Test",
    fullName: "Alice Brown",
  },
];
