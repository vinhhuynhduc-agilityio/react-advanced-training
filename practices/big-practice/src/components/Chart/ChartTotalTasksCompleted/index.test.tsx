import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

// component
import { ChartTotalTasksCompleted } from '.';

// mock data
import { mockTasks } from '@/mocks';

// Mock AgCharts
jest.mock('ag-charts-react', () => ({
  AgCharts: () => <div data-testid="mocked-chart"></div>,
}));

// Mock helpers
jest.mock('@/components/Chart/helpers', () => ({
  ...jest.requireActual('@/components/Chart/helpers'),
  formatDataForChartTotalTasks: jest.fn(() => [
    { month: 'January', '2023': 5, '2024': 10 },
    { month: 'February', '2023': 3, '2024': 7 },
  ]),
}));

describe('ChartTotalTasksCompleted Component', () => {
  it('renders the mocked chart when isLoading is false', () => {
    render(
      <ChartTotalTasksCompleted isLoading={false} tasks={mockTasks} />
    );

    const mockedChart = screen.getByTestId('mocked-chart');
    expect(mockedChart).toBeInTheDocument();
    expect(screen.getByRole('figure', { name: /chart showing total tasks completed/i })).toBeInTheDocument();
  });

  it('matches snapshot for mocked chart', () => {
    const { container } = render(
      <ChartTotalTasksCompleted isLoading={false} tasks={mockTasks} />
    );

    expect(container).toMatchSnapshot();
  });

  it('renders a spinner when isLoading is true', () => {
    render(
      <ChartTotalTasksCompleted isLoading={true} tasks={mockTasks} />
    );

    const spinner = screen.getByRole('figure', {
      name: /loading chart: total tasks completed/i,
    });
    expect(spinner).toBeInTheDocument();

    const mockedChart = screen.queryByTestId('mocked-chart');
    expect(mockedChart).not.toBeInTheDocument();
  });

  it('formats data correctly using the helper function', () => {
    render(
      <ChartTotalTasksCompleted isLoading={false} tasks={mockTasks} />
    );

    // Check if the mocked helper function was called
    const { formatDataForChartTotalTasks } = jest.requireMock('@/components/Chart/helpers');
    expect(formatDataForChartTotalTasks).toHaveBeenCalledWith(mockTasks);
  });
});
