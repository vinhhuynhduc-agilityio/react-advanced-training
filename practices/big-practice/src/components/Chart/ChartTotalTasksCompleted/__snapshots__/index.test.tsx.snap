// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartTotalTasksCompleted Component matches snapshot for mocked chart 1`] = `
<div>
  <div
    aria-label="Chart showing total tasks completed"
    class="flex-1 bg-white border border-customBorder h-[302px]"
    id="chart-total-tasks-completed"
    role="figure"
  >
    <div
      data-testid="mocked-chart"
    />
  </div>
</div>
`;
